architecture/index.md

```diff
@@ -39,6 +39,55 @@ API resources in the second row (PipelineRun, Snapshot) should primarilly be tho
 data-plane resources. The system responds to user requests by creating and managing the lifecycle of
 these resources.
 
+- An [Application] represents a functionally coherent set of [Components] that should be built,
+  tested, and released together. The user provides and names their [Applications]. They are
+  generally long-lived and don't change much after they are created.
+- A [Component] represents an OCI artifact to build from a particular git branch (and a particular context
+  directory). It is owned by an [Application]. The user provides and names their [Components] and
+  specifies which git repositories and branches define how the artifact is built. The user may add,
+  remove, or change [Components] over the lifespan of an [Application].
+- A [Snapshot] represents a collection of particular OCI artifact, mapped to the [Components] of an
+  [Application]. A [Snapshot] is owned by an [Application]. Generally, the [integration-service]
+  produces new [Snapshots] automatically in response to completed builds. It is also possible for
+  the user to construct and provide [Snapshots] to test or release deliberate combinations of
+  OCI artifacts. Old [Snapshots] are garbage collected if not bound to other resources like
+  a [SnapshotEnvironmentBinding] or a [Release].
+- An [IntegrationTestScenario] represents a test that should be run against new [Snapshots] that
+  belong to a particular [Application]. It is owned by an [Application]. The user provides writes
+  tekton pipelines that test their application and registers them with the system by creating
+  [IntegrationTestScenarios].
+- An [Environment] represents a destination that [application-service] and the [gitops-service] can
+  deploy to. Multiple [Applications] in the same workspace can deploy to the same
+  [Environment]. An [Environment] may be long lived (like the `development` [Environment] provided
+  by the system as part of the initialization of a user's workspace) or may be short-lived and
+  ephemeral (like the temporary [Environments] created by the [integration-service] as a part of its
+  test execution processes).
+- A [DeploymentTargetClaim] represents a request to provision a new deployment target. An
+  [Environment] is backed by a [DeploymentTargetClaim]. The [DeploymentTargetClaim] causes new
+  deployment targets to be provisioned which are represented by [DeploymentTargets] that get bound
+  to it. These are typically created and destroyed by the [integration-service] as a part of its
+  test execution processes.
+- A [SnapshotEnvironmentBinding] represents a request to deploy a particular set of OCI artifacts
+  (represented by a [Snapshot]) to a particular location (represented by an [Environment]).
+  A [SnapshotEnvironmentBinding] is owned by an [Application]. The [integration-service promotes
+  OCI artifacts] by updating the [SnapshotEnvironmentBinding] associated with an [Environment]. The
```

Comment:
This seems to imply that Snapshots are also long-lived, and only updated as new builds of Components are created? Or, is the binding updated with new Snapshot instances?

Reply:
Snapshots are long lived in that the CRs are not deleted. Some workspaces with more builds have hit the CR limits and need to be manually cleaned up. We are implementing a periodic garbage collection cron until proper CR storage can be implemented (i.e. considering something like https://github.com/adambkaplan/kubearchive).

When a new Component build is created, a Snapshot is created for testing. There is a [two-phase approach](https://github.com/redhat-appstudio/architecture/blob/main/ADR/0015-integration-service-two-phase-architecture.md) for Snapshot promotion based on the success of the required IntegrationTestScenarios.

SEBs are updated with new Snapshot CRs after successfully passing the Composite promotion phase.

Reply:
Ah - a missing important detail is that Snapshots are *immutable*! Updated in ad2961c.

Reply:
Immutable except for their statuses.